### PR TITLE
Graceful shutdown feature implementation

### PR DESCRIPTION
**Pull request by gfigiel**
Solution for Issue #75 Graceful shutdown feature implementation

Based on branch 2.x

### CODE REVIEW

**Comment by deruelle** [comment]
Thanks @gfigiel ! @SergeyLee can you review ?

**Comment by gfigiel** [comment]
Hi,
Please see the implementaiton for CAP RA as an example for using new interface:
https://github.com/ProIDS/jain-slee.ss7/tree/issue%2314-graceful-shutdown

**Comment by SergeyLee** [comment]
Hi @gfigiel 
Who did testing JSLEE Release with Graceful shutdown on your side?

**Comment by gfigiel** [comment]
Hi Sergey - the slee in our TB is now ready to be tested but we have technical issues in test environment that blocked tests already. I assume the testing will be finished till end of this week by me and Dominik (@dblacha).

Br,
Greg

**Comment by SergeyLee** [comment]
Hi @gfigiel I have reviewed your commits and have only one question: can you create also non regression tests for this feature?

**Comment by gfigiel** [comment]
Hi @SergeyLee 
We have still not finished new release tests as jDiameter extension problems are now "on air".
Reagrding your question - please clarify what kind of tests do you mean? We reviewed the jain-slee but we have not found unit tests there.

**Comment by SergeyLee** [comment]
Hi @gfigiel 
JAIN SLEE has not unit tests for all modules and all purposes, you are right. But for some of them we have them. Like for components, congestion, jmx-property-editors, transactions, etc.
For Graceful shutdown feature these tests is needed in the resource module, because main changes was created there.

File: container/spi/src/main/java/org/mobicents/slee/container/resource/GracefullyStopableResourceAdaptor.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+package org.mobicents.slee.container.resource;
+
+/**
+ * Interface to be implemented by RA objects that support SLEE Graceful Shutdown feature.
+ *
+ *
+ * @author <a href="mailto:grzegorz.figiel@pro-ids.com"> Grzegorz Figiel (ProIDS sp. z o.o.)</a>
+ *
+ */
+public interface GracefullyStopableResourceAdaptor {
```
Comment (jaimecasero) [inline container/spi/src/main/java/org/mobicents/slee/container/resource/GracefullyStopableResourceAdaptor.java]: so,every resource adaptor needs to implement the graceful itnerface to comply with expected behavior...
I thought this was more a jslee container issue, that didnt affected RA level...

File: container/spi/src/main/java/org/mobicents/slee/container/resource/ResourceAdaptorObjectState.java [hunk 2]
```diff
@@ -24,10 +24,11 @@
...
  * The states for the Resource Adaptor state machine (see SLEE 1.1 specs 15.7.1)
  * @author martins
+ * @author <a href="mailto:grzegorz.figiel@pro-ids.com"> Grzegorz Figiel (ProIDS sp. z o.o.)</a>
  *
  */
 public enum ResourceAdaptorObjectState {
 
-	UNCONFIGURED, INACTIVE, ACTIVE, STOPPING
-	
+	UNCONFIGURED, INACTIVE, ACTIVE, STOPPING, STOPPING_GRACEFULLY
```
Comment (jaimecasero) [inline container/spi/src/main/java/org/mobicents/slee/container/resource/ResourceAdaptorObjectState.java]: i can see the FSM needs to add a new state... what is the main difference in terms of state transitions between STOPPING and STOPPING_GRACEFULLY?

File: container/common/src/main/java/org/mobicents/slee/container/management/jmx/SleeManagementMBeanImpl.java [hunk 3]
```diff
@@ -220,19 +237,134 @@ private void notifyStateChange(SleeState oldState, SleeState newState) {
...
+
+		return "Container Graceful Shutdown requested successfully.";
+	}
+
+	/**
+	 * Stop the SLEE. Should do it in a non-blocking manner.
 	 * 
 	 * @see javax.slee.management.SleeManagementMBean#stop()
 	 */
 	public void stop() throws InvalidStateException, ManagementException {
```
Comment (jaimecasero) [inline container/common/src/main/java/org/mobicents/slee/container/management/jmx/SleeManagementMBeanImpl.java]: why do we need a nonGraceful stop? As stated in Spec all stops should be gracefull, entering the STOPPING state, and causing all RAs to enter stopping state as well...14.5.1 Operational states of the SLEE

**Comment by satanatoly** [comment]
implemented other version of graceful shutdown in commercial version, closing

**Pull request closed without merging**